README.md
# Face++
## Face recognition for Homey ##
Use snapshots from a camera connected to Homey, and let Face++ do AI based face recognition. For every face the app detects, a flow will be triggered. Homey will automatically try to match the faces to a set of known faces that you can upload yourself. The app will also estimate gender, age, emotion, and whether the person is wearing (sun)glasses or a mask.

<img src="https://aws1.discourse-cdn.com/business4/uploads/athom/original/3X/2/4/249732c22df6c415cd2e82ce50477b050ab799c7.jpeg" alt="Face++" width="300">

## Use cases ##
* Presence detection
* Security
* Face based access control
* Fun

## Privacy ##
The Homey app uses a cloud based face recognition service called Face++. Using the Face++ service has severe privacy implications. Use at own risk. Read and understand your local privacy legislation, and the terms of use from Face++.

## Performance ##
The results depend very much on the quality of the snapshots. A face will be best recognized if the lighting is good and when the face is frontal with a width of at least 75 pixels. In practice it means you need to have good control over when the snapshot is taken in a videostream. The free Face++ service allows you to upload around 1 or 2 images per second. The face detection usually is ready within 1 to 3 seconds from uploading an image (via an action flowcard).

## Test image ##
To make testing your flow easier, a test image is available as global image token. It has five faces in it.

<img src="https://aws1.discourse-cdn.com/business4/uploads/athom/original/3X/f/e/fe69c1e5aeaf6e42850f55f89064f07d58e6bd64.jpeg" alt="Face++" width="300">




changelog.txt
v2.0.2  25.04.2023 jimp@0.22.7. Partial fix Homey2023 compatibility.
v2.0.1  02.02.2023 Fixed Homey2023 compatibility.
v2.0.0  31.01.2023 Rewrite to SDK3. jimp@0.16.2
v1.2.2  24.05.2021 Fix HTTP(S) timeout. Improved logger. jimp@0.16.1
v1.2.1  28.06.2020 Added cropped face image tag in trigger card. Small fixes.
v1.1.7  07.06.2020 Fix INVALID_OUTER_ID when no face added. Fix THRESHOLD error after clean install.
v1.1.6  11.05.2020 Fix not able to send to Telegram on NO_MATCH. Fix app settings pop-up messages.
v1.1.5  05.05.2020 Stability improvements. Added toggle API Secret view in app settings. Fix threshold settings.
v1.1.4  03.05.2020 New settings and face set scheme. Minor fixes.
v1.1.2  03.05.2020 Fix not able to add faces. Add warning that each Homey needs own API key.
v1.1.1  03.05.2020 Added clickable registration link. Added log filters. Fixed flowtoken_already_exists. Fixed ENOENT: no such file or directory.
v1.1.0  02.05.2020 Added cloud queueing and rate limiting.
v1.0.4  01.05.2020 Added test image. Fixed unhandledRejection errors. Added API check. Improved app logs. Removed INVALID_OUTER_ID log.
v1.0.3  29.04.2020 App init and frontend bugfixes.
v1.0.2  29.04.2020 Homey Appstore compliancy.
v1.0.1  29.04.2020 Several small fixes.
v1.0.0  28.04.2020 First release.

captureLogs.js
/*
Copyright 2020, Robin de Gruijter (gruijter@hotmail.com)

This file is part of com.gruijter.facepp.

com.gruijter.facepp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

com.gruijter.facepp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with com.gruijter.facepp.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

const Homey = require('homey');
const StdOutFixture = require('fixture-stdout');
const fs = require('fs');

class captureLogs {
	// Log object to keep logs in memory and in persistent storage
	// captures and reroutes Homey's this.log (stdout) and this.err (stderr)

	constructor(logName, logLength) {
		this.logName = logName || 'log';
		this.logLength = logLength || 50;
		this.logFile = `/userdata/${this.logName}.json`;
		this.logArray = [];
		this.getLogs();
		this.captureStdOut();
		this.captureStdErr();
	}

	getLogs() {
		try {
			const log = fs.readFileSync(this.logFile, 'utf8');
			this.logArray = JSON.parse(log);
			Homey.app.log('logfile retrieved');
			return this.logArray;
		} catch (error) {
			if (error.message.includes('ENOENT')) return [];
			Homey.app.error('error parsing logfile: ', error.message);
			return [];
		}
	}

	saveLogs() {
		try {
			fs.writeFileSync(this.logFile, JSON.stringify(this.logArray));
			Homey.app.log('logfile saved');
			return true;
		} catch (error) {
			Homey.app.error('error writing logfile: ', error.message);
			return false;
		}
	}

	deleteLogs() {
		try {
			fs.unlinkSync(this.logFile);
			this.logArray = [];
			Homey.app.log('logfile deleted');
			return true;
		} catch (error) {
			if (error.message.includes('ENOENT')) return false;
			Homey.app.error('error deleting logfile: ', error.message);
			return false;
		}
	}

	captureStdOut() {
		// Capture all writes to stdout (e.g. this.log)
		this.captureStdout = new StdOutFixture({ stream: process.stdout });
		this.captureStdout.capture((string) => {
			if (this.logArray.length >= this.logLength) {
				this.logArray.shift();
			}
			this.logArray.push(string);
			// return false;	// prevent the write to the original stream
		});
		Homey.app.log('capturing stdout');
		// captureStdout.release();
	}

	captureStdErr() {
		// Capture all writes to stderr (e.g. this.error)
		this.captureStderr = new StdOutFixture({ stream: process.stderr });
		this.captureStderr.capture((string) => {
			if (this.logArray.length >= this.logLength) {
				this.logArray.shift();
			}
			this.logArray.push(string);
			// return false;	// prevent the write to the original stream
		});
		Homey.app.log('capturing stderr');
		// captureStderr.release();
	}

	releaseStdOut() {
		this.captureStdout.release();
	}

	releaseStdErr() {
		this.captureStderr.release();
	}

}

module.exports = captureLogs;


settings/settings.js
/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

// tab 2 stuff here
function displayLogs(lines) {
	$('#loglines').html(lines);
}

function updateLogs() {
	try {
		displayLogs('');
		const showLogs = $('#show_logs').prop('checked');
		const showErrors = $('#show_errors').prop('checked');
		const showNoDetects = $('#show_no_detections').prop('checked');
		Homey.api('GET', 'getlogs/', null, (err, result) => {
			if (!err) {
				let lines = '';
				result
					.reverse()
					.forEach((line) => {
						if (!showLogs) {
							if (line.includes('[log]')) return;
						}
						if (!showErrors) {
							if (line.includes('[err]')) return;
						}
						if (!showNoDetects) {
							if (line.includes(' found in image')) return;
						}
						const logLine = line.replace(' [FaceApp]', '');
						lines += `${logLine}<br />`;

					});
				displayLogs(lines);
			} else {
				displayLogs(err);
			}
		});
	} catch (e) {
		displayLogs(e);
	}
}

function deleteLogs() {
	Homey.confirm(Homey.__('settings.tab2.deleteWarning'), 'warning', (error, result) => {
		if (result) {
			Homey.api('GET', 'deletelogs/', null, (err) => {
				if (err) {
					Homey.alert(err.message, 'error'); // [, String icon], Function callback )
				} else {
					Homey.alert(Homey.__('settings.tab2.deleted'), 'info');
					updateLogs();
				}
			});
		}
	});
}

// tab 3 stuff here
function checkKeys() {
	Homey.get('settingsKey', (error, set) => {
		if (error || !set) return Homey.alert('API Key and Secret are not saved!', 'error');
		if (!set.apiKey || set.apiKey.length < 10
			|| !set.apiSecret || set.apiSecret.length < 10) return Homey.alert('API Key and Secret are not correctly saved!', 'error');
		return true;
	});
}

function showInfo3() {
	const x = document.getElementById('apiSecret');
	x.type = 'password';
	Homey.get('settingsKey', (err, set) => {
		if (err || !set) return;
		$('#apiKey').val(set.apiKey);
		$('#apiSecret').val(set.apiSecret);
	});
	Homey.get('settingsMatch', (err, set) => {
		if (err || !set) return;
		$('#threshold').val(set.threshold);
	});
}

function togglePasswordView() {
	const x = document.getElementById('apiSecret');
	if (x.type === 'password') {
		x.type = 'text';
	} else {
		x.type = 'password';
	}
}

async function saveSettingsKey() {
	try {
		const apiKey = $('#apiKey').val();
		const apiSecret = $('#apiSecret').val();
		await Homey.set('settingsKey', { apiKey, apiSecret });
		return Homey.alert('API Keys are saved!', 'info');
	} catch (error) {
		return Homey.alert(error.message, 'error');
	}
}

async function saveSettingsMatch() {
	try {
		const threshold = $('#threshold').val();
		if (threshold < 0 || threshold > 100) throw Error('Threshold must be between 0 and 100');
		await Homey.set('settingsMatch', { threshold });
		return Homey.alert('Settings are saved!', 'info');
	} catch (error) {
		return Homey.alert(error.message, 'error');
	}
}

// tab 4 stuff here
let faceSet = {};
let faceInfo = {};

function getFaceSet() {
	return new Promise((resolve, reject) => {
		Homey.get('face_set', (err, set) => {
			if (err) return reject(err);
			faceSet = set || {};
			return resolve(faceSet);
		});
	});
}

function clearInfo() {
	// $('#label').val('');
	$('#faceToken').val('');
	$('#facequality').val(0);
	$('#gender').val('');
	$('#age').val(0);
	$('#glass').val('');
}

function showInfo(face, detected) {
	// add new face; disable save, enable photo upload, clear label
	if (!face || !face.face_token) {
		$('#label').val('');
		$('#file').show();
		$('#preview').prop('src', '../assets/icon.svg');
		$('#preview').prop('src', imgPath);
		$('#saveFace').attr('disabled', true);
		$('#deleteFace').attr('disabled', true);
		clearInfo();
		return;
	}
	// newly detected, unsaved face; enable save, enable upload photo
	if (detected) {
		$('#file').show();
		$('#saveFace').attr('disabled', false);
		$('#deleteFace').attr('disabled', true);
	}
	// existing face: load image, load label, disable upload photo
	if (!detected) {
		$('#file').hide();
		$('#saveFace').attr('disabled', false);
		$('#deleteFace').attr('disabled', false);
		$('#label').val(face.label);
		const imgPath = `../userdata/${face.face_token}.img`;
		$('#preview').prop('src', imgPath);
	}
	$('#faceToken').val(face.face_token);
	$('#facequality').val(face.attributes.facequality.value);
	$('#gender').val(face.attributes.gender.value);
	$('#age').val(face.attributes.age.value);
	$('#glass').val(face.attributes.glass.value);
}

function faceSelected() {
	const selectedFace = $('#faceList').val();
	// if (selectedFace === ('null' || 'undefined')) clearInfo();
	faceInfo = faceSet[selectedFace] || {};
	showInfo(faceInfo);
}

// function to populate the dropdown list
async function fillDropdown() {
	// first empty the dropdownlist
	const dropDown = document.getElementById('faceList'); // $('#faceList');
	while (dropDown.length > 0) {
		dropDown.remove(dropDown.length - 1);
	}
	// now fill the dropdown list and add an empty item in the dropdown list
	const newOption = document.createElement('option');
	newOption.text = __('settings.tab4.newFace');
	newOption.value = null;
	dropDown.add(newOption);
	faceSet = await getFaceSet();
	Object.keys(faceSet).forEach((key) => {
		const option = document.createElement('option');
		option.text = `${faceSet[key].label}-${faceSet[key].face_token}`;
		option.value = faceSet[key].face_token;
		dropDown.add(option);
	});
	faceSelected();
}

// detect face info
async function detectFace(img) {
	try {
		const faces = await new Promise((resolve, reject) => {
			Homey.api('POST', 'detectface/', { img }, (err, result) => {
				clearInfo();
				if (err) return reject(err);
				if (!result.face_num) return reject(Error('No face detected'));
				if (result.face_num > 1) return reject(Error(`${result.face_num} faces detected, but only one allowed`));
				return resolve(result);
			});
		});
		faceInfo = faces.faces[0] || {};
		showInfo(faceInfo, true);
		return faces;
	} catch (error) {
		return Homey.alert(error.message, 'error');
	}
}

// load image file
function loadFile(event) {
	const imgFile = event.target.files[0];
	const preview = document.getElementById('preview');
	// make base64 file and display
	const reader = new FileReader();
	reader.readAsDataURL(imgFile); // converts the blob to base64 and calls onload
	reader.onload = () => {
		const imgBase64 = reader.result; // data url
		preview.src = imgBase64; // show in frontend
		clearInfo();
		if (imgBase64.length > 1048000) return Homey.alert('Image size is too large', 'error');
		$('#faceToken').val('Analyzing. Hang on....');
		// Homey.alert('Face is being analyzed. Hang on...', 'info');
		return detectFace(imgBase64);
	};
}

function addFace() {
	// const img = $('#preview').attr('src');
	const img = document.getElementById('preview').src;
	return Homey.api('POST', 'addface/', { img, faceInfo }, (err, result) => {
		if (err) {
			Homey.alert(err.message, 'error'); // [, String icon], Function callback )
			return;
		}
		fillDropdown();
		Homey.alert('Face is saved!', 'info');
	});
}

async function updateFace() {
	try {
		const fset = await getFaceSet();
		fset[faceInfo.face_token] = faceInfo;
		await Homey.set('face_set', fset);
		Homey.alert('Face label is updated!', 'info');
	} catch (error) {
		Homey.alert(error.message, 'error');
	}
}

function saveFace() {
	faceInfo.label = $('#label').val();
	faceInfo.face_token = $('#faceToken').val();
	if (!faceInfo.face_token || faceInfo.face_token.length !== 32) return Homey.alert('Invalid face token. Upload a valid face image.', 'error');
	// check if new face, or only updated label
	if (Object.prototype.hasOwnProperty.call(faceSet, faceInfo.face_token)) {
		return updateFace();
	}
	return addFace();
}

function deleteFace() {
	const selectedFace = $('#faceList').val();
	if (selectedFace === ('null' || 'undefined')) return;
	Homey.api('POST', 'deleteface/', { faceInfo: { face_token: selectedFace } }, (err, result) => {
		if (err) {
			Homey.alert(err.message, 'error'); // [, String icon], Function callback )
			return;
		}
		fillDropdown();
		Homey.alert('Face is deleted!', 'info');
	});
}

// generic stuff here
function showTab(tab) {
	if (tab === 2) updateLogs();
	if (tab === 3) {
		checkKeys();
		showInfo3();
	}
	if (tab === 4) {
		checkKeys();
		fillDropdown();
	}
	$('.tab').removeClass('tab-active');
	$('.tab').addClass('tab-inactive');
	$(`#tabb${tab}`).removeClass('tab-inactive');
	$(`#tabb${tab}`).addClass('active');
	$('.panel').hide();
	$(`#tab${tab}`).show();
}

function onHomeyReady(homeyReady) {
	Homey = homeyReady;
	showInfo3();
	showTab(1);
	Homey.ready();
}


app.js
/*
Copyright 2020 - 2023, Robin de Gruijter (gruijter@hotmail.com)

This file is part of com.gruijter.facepp.

com.gruijter.facepp is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

com.gruijter.facepp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with com.gruijter.facepp.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

const Homey = require('homey');
const fs = require('fs');
const { networkInterfaces } = require('os');
// const { Duplex } = require('stream');
const { Readable } = require('stream');
const Jimp = require('jimp');
// const util = require('util');
const Logger = require('./captureLogs');
const FacePP = require('./facepp');

// const setTimeoutPromise = util.promisify(setTimeout);

// convert buffer to/from base64 encoded string
const base64Encode = (img) => Buffer.from(img).toString('base64');
const base64Decode = (base64) => Buffer.from(base64, 'base64');

const bufferToStream = (buffer) => {
	const stream = Readable.from(buffer);
	return stream;
};

const streamToBuffer = (stream) => new Promise((resolve, reject) => {
	const buffers = [];
	stream.on('error', reject);
	stream.on('data', (data) => buffers.push(data));
	stream.on('end', () => resolve(Buffer.concat(buffers)));
});

// image crop
const crop = async (imgBuffer, region) => {	// returns image buffer
	try {
		const image = await Jimp.read(imgBuffer);
		image.crop(region.left, region.top, region.width, region.height);
		return image.getBufferAsync(image.getMIME());
	} catch (error) {
		return error;
	}
};

const getEmotion = (emotions) => {
	const emotionsArray = Object.keys(emotions).map((emotion) => ({ emotion, value: emotions[emotion] }));
	if (!emotionsArray || !emotionsArray[0]) return 'undefined';
	return emotionsArray.sort((a, b) => b.value - a.value)[0].emotion;
};

class FaceApp extends Homey.App {

	async onInit() {
		try {
			if (!this.logger) this.logger = new Logger({ name: 'log', length: 200, homey: this.homey });

			process.on('unhandledRejection', (error) => {
				this.error('unhandledRejection! ', error);
			});
			process.on('uncaughtException', (error) => {
				this.error('uncaughtException! ', error);
			});
			this.homey
				.on('unload', () => {
					this.log('app unload called');
					// save logs to persistant storage
					this.logger.saveLogs();
				})
				.on('cpuwarn', () => {
					this.log('cpuwarn!');
				})
				.on('memwarn', () => {
					this.log('memwarn!');
				});

			// first remove settings events listener
			if (this.listeners && this.listeners.set) {
				this.homey.settings.removeListener('set', this.listeners.set);
			}

			// listen to change settings events
			this.listeners.set = (key) => {
				if (key === 'settingsMatch') {
					this.log('Face Match settings changed:', JSON.stringify(this.homey.settings.get('settingsMatch')));
				}
				if (key === 'settingsKey') {
					this.log('API Key settings changed, reloading app now');
				}
				this.logger.saveLogs();
				this.log('Re-initing app now');
				this.onInit();
			};
			this.homey.settings
				.on('set', this.listeners.set);

			this.matchSettings = this.homey.settings.get('settingsMatch') || { threshold: 75 }; // threshold
			const { apiKey, apiSecret } = this.homey.settings.get('settingsKey') || {}; // apiKey, apiSecret
			if (!apiKey || !apiSecret) {
				this.log('No API key entered in app settings');
				return;
			}

			const interfaces = networkInterfaces();
			let mac = null;

			if (interfaces.eth1 && interfaces.eth1[0].address) {
				mac = interfaces.eth1[0].mac;
			} else if (interfaces.wlan0 && interfaces.wlan0[0].address) {
				mac = interfaces.wlan0[0].mac;
			} else if (interfaces.eth0 && interfaces.eth0[0].address) {
				mac = interfaces.eth0[0].mac;
			}
			if (!mac) throw Error('No active network connection found.');

			this.outerID = `homey_${mac}`;
			const options = {
				key: apiKey,
				secret: apiSecret,
			};
			this.FAPI = new FacePP(options);

			// register flows and tokens
			await this.registerFlowCards();
			await this.registerFlowTokens();

			// sync Face++ set and local .img files
			await this.syncFaceset();

			this.log('Face++ is running...');

			// do garbage collection every 10 minutes
			// this.intervalIdGc = setInterval(() => {	global.gc(); }, 1000 * 60 * 10);

		} catch (error) {
			this.error(error);
		}

	}

	//  stuff for frontend API
	deleteLogs() {
		return this.logger.deleteLogs();
	}

	getLogs() {
		return this.logger.logArray;
	}

	// returns an array of detected scenes and objects
	async detectObjects(imgBase64) {
		try {
			const options = {
				image_base64: imgBase64,
			};
			const result = await this.FAPI.detectObjects(options);
			return Promise.resolve(result);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// returns a formatted array of recognised objects, and triggers flowCards
	async searchObjects(imgBase64, origin) {
		try {
			const { scenes, objects } = await this.detectObjects(imgBase64);

			const scene = {
				value: 'undefined',
				confidence: 0,
			};
			if (scenes.length > 0) {
				const bestScene = scenes.sort((a, b) => b.confidence - a.confidence)[0];
				scene.value = bestScene.value;
				scene.confidence = bestScene.confidence;
			}
			const objectResult = objects.map(async (obj) => {
				const tokens = {
					origin: origin || 'undefined',
					scene: scene.value,
					scene_confidence: scene.confidence,
					object: obj.value,
					object_confidence: obj.confidence,
				};
				this.log(tokens);
				// this.flows.objectDetectedTrigger.trigger(tokens);
				return tokens;
			});
			await Promise.all(objectResult);
			if (objectResult.length === 0) this.log(`no objects found in image ${origin}`);
			return objectResult;
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// returns an array of detected faces plus attributes
	async detectFaces(imgBase64) {
		try {
			const options = {
				image_base64: imgBase64,
			};
			const result = await this.FAPI.detectFaces(options);
			return Promise.resolve(result);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// get faceImageToken Stream
	async getFaceToken(imgBuffer, bounds) {
		try {
			const croppedFace = await crop(imgBuffer, bounds);
			const imgFunct = async (stream) => {
				const imgStream = bufferToStream(croppedFace);
				imgStream.pipe(stream);
			};
			const faceImage = await this.homey.images.createImage();
			faceImage.setStream(imgFunct);
			setTimeout(() => { faceImage.unregister(); }, 1000 * 60 * 3); // image stream is available for 3 minutes
			return Promise.resolve(faceImage);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// returns a formatted array of recognised faces, and triggers flowCards
	async searchFaces(imgStream, origin) {
		try {
			const imgBuffer = await streamToBuffer(imgStream);
			const imgBase64 = base64Encode(imgBuffer);
			const { faces } = await this.detectFaces(imgBase64);
			const homeySet = this.homey.settings.get('face_set') || {};
			const searchResult = faces.map(async (face) => {
				const snapshot = await this.getFaceToken(imgBuffer, face.face_rectangle);
				const tokens = {
					origin: origin || 'undefined',
					label: 'NOMATCH',
					confidence: 0,
					token: face.face_token,
					quality: face.attributes.facequality.value,
					gender: face.attributes.gender.value,
					age: face.attributes.age.value,
					emotion: getEmotion(face.attributes.emotion),
					normalglass: face.attributes.glass.value === 'Normal',
					sunglass: face.attributes.glass.value === 'Dark',
					mask: face.attributes.mouthstatus.surgical_mask_or_respirator.value > 50,
					face_image_token: snapshot,
				};
				if (Object.keys(homeySet).length > 0) {
					const options = {
						outer_id: this.outerID,
						face_token: face.face_token,
					};
					const result = await this.FAPI.searchFaces(options);
					const { threshold } = this.matchSettings;
					if (result.results) {
						const bestMatch = result.results.sort((a, b) => b.confidence - a.confidence)[0];
						if (bestMatch.confidence > threshold) {
							tokens.label = homeySet[bestMatch.face_token].label;
							tokens.confidence = bestMatch.confidence;
							tokens.token = homeySet[bestMatch.face_token].face_token;
						}
					}
				}

				const logTokens = { ...tokens };
				logTokens.face_image_token = tokens.face_image_token.cloudUrl;
				this.log(logTokens);

				const faceDetectedTrigger = this.homey.flow.getTriggerCard('face_detected');
				faceDetectedTrigger.trigger(tokens);
				return tokens;
			});
			await Promise.all(searchResult);
			if (searchResult.length === 0) this.log(`no faces found in image ${origin}`);
		} catch (error) {
			this.error(error);
		}
	}

	async analyzeImage(args) {
		try {
			// get the contents of the image
			const image = args.droptoken;
			if (!image || !image.getStream) throw Error('no valid image stream');
			// if (typeof image === 'undefined' || image == null) return false;
			const imageStream = await image.getStream();
			this.log('analyzing image from face', args.origin);
			await this.searchFaces(imageStream, args.origin);
		} catch (error) {
			this.error(error);
		}
	}

	// sync the Homey set with Face++ set and userdata/*.img
	async syncFaceset() {
		try {
			// migrating from old sceme
			if (this.migrating) {
				this.log('migrating Face++ face set to new scheme');
				const options = {
					outer_id: 'homey',
					face_tokens: 'RemoveAllFaceTokens',
				};
				await this.FAPI.fsRemoveFace(options)
					.catch((error) => {
						if (error.message.includes('INVALID_OUTER_ID')) return;
						this.error(error);
					});
				this.migrating = false;
			}

			const homeySet = this.homey.settings.get('face_set') || {};
			// remove all face tokens in Face++ set
			const options = {
				outer_id: this.outerID,
				face_tokens: 'RemoveAllFaceTokens',
			};
			await this.FAPI.fsRemoveFace(options)
				.catch((error) => {
					if (error.message.includes('INVALID_OUTER_ID')) return;
					this.error(error);
				});

			// add all face tokens from Homey set to Face++ set
			const ready = Object.keys(homeySet).map((face) => {
				const opts = {
					outer_id: this.outerID,
					face_tokens: face,
				};
				return this.FAPI.fsAddFace(opts);
			});
			await Promise.all(ready);

			// purge all img files that are not in homeySet
			const files = fs.readdirSync('/userdata');
			files.forEach((file) => {
				if (file.includes('.img')) {
					const token = file.split('.')[0];
					if (!Object.keys(homeySet).includes(token)) {
						fs.unlinkSync(`/userdata/${file}`);
					}
				}
			});
			this.log('ready syncing face set with cloud');
		} catch (error) {
			this.error('CRITICAL: FACE SET CLOUD SYNC FAILED!');
			this.error(error);
		}
	}

	// add face to faceset
	async addFace(body) {
		try {
			if (!body.faceInfo || !body.faceInfo.face_token) throw Error('No Face Token; cannot add face');

			// save base64 file as jpeg/img
			const filename = `/userdata/${body.faceInfo.face_token}.img`;
			const file64 = body.img.replace(/^data:image\/(\w+);base64,/, '');
			const file = base64Decode(file64);
			fs.writeFileSync(filename, file);

			// add face to set in Face++
			const options = {
				outer_id: this.outerID,
				face_tokens: body.faceInfo.face_token,
			};
			await this.FAPI.fsAddFace(options);

			// store face persistent in Homey
			const faceSet = this.homey.settings.get('face_set') || {};
			faceSet[body.faceInfo.face_token] = body.faceInfo;
			this.homey.settings.set('face_set', faceSet);
			this.log('added face', options);
			return Promise.resolve(true);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// delete face from faceset
	async deleteFace(body) {
		try {
			// remove img file
			const filename = `/userdata/${body.faceInfo.face_token}.img`;
			if (fs.existsSync(filename)) fs.unlinkSync(filename);

			// remove face from in Homey
			const faceSet = this.homey.settings.get('face_set') || {};
			delete faceSet[body.faceInfo.face_token];
			this.homey.settings.set('face_set', faceSet);

			// delete face from set in Face++
			const options = {
				outer_id: this.outerID,
				face_tokens: body.faceInfo.face_token,
			};
			await this.FAPI.fsRemoveFace(options);
			this.log('deleted face', options);
			return Promise.resolve(true);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// register flow cards
	async registerFlowCards() {
		try {
			if (this.hasFlows) return;
			// action cards
			const analyzeImageAction = this.homey.flow.getActionCard('search_faces');
			analyzeImageAction.registerRunListener((args) => this.analyzeImage(args));
			this.hasFlows = true;
		} catch (error) {
			this.error(error);
		}
	}

	// register global flow tokens
	async registerFlowTokens() {
		try {
			// unregister tokens first
			if (!this.tokens) this.tokens = {};
			const ready = Object.keys(this.tokens).map((token) => Promise.resolve(this.homey.ManagerFlow.unregisterToken(this.tokens[token])));
			await Promise.all(ready);

			// register the test image
			const fileName = this.homey.platformVersion === 2 ? '/app/assets/images/test.jpg' : '/assets/images/test.jpg';
			const testImage = await this.homey.images.createImage();
			testImage.setPath(fileName);

			// create a droptoken for test image
			this.tokens.testImageToken = await this.homey.flow.createToken('test_image_token', {
				type: 'image',
				title: 'Face++ Test Image',
			});
			await this.tokens.testImageToken.setValue(testImage);

			return Promise.resolve(this.tokens);
		} catch (error) {
			return Promise.resolve(error);
		}
	}

}

module.exports = FaceApp;

/*
  {
    face_token: '2485daa964b877b1a6c6b109f409b561',
    face_rectangle: { top: 173, left: 218, width: 87, height: 87 },
    attributes: {
      gender: [Object],
      age: [Object],
      smile: [Object],
      eyestatus: [Object],
      emotion: [Object],
      facequality: [Object],
      beauty: [Object],
      mouthstatus: [Object],
      eyegaze: [Object],
      skinstatus: [Object],
      glass: [Object]
    }
  },

  face_image_token: Image {
    id: '75aa0755-e6de-4868-a4dd-9d16feb259a4',
    __client: { emit: [Function: bound __onImageEmit] },
    _type: 'buffer',
    _source: <Buffer ff d8 ff e0 00 10 4a 46 49 46 00 01 01 00 00 01 00 01 00 00 ff db 00 84 00 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 01 ... 7556 more bytes>,
    cloudUrl: 'https://5c4968afb1b00f14323059f6.connect.athom.com/image/75aa0755-e6de-4868-a4dd-9d16feb259a4/image',
    localUrl: 'http://10.0.0.61/image/75aa0755-e6de-4868-a4dd-9d16feb259a4/image'
  }

  PassThrough {
  _readableState: ReadableState {
    objectMode: false,
    highWaterMark: 16384,
    buffer: BufferList { head: null, tail: null, length: 0 },
    length: 0,
    pipes: null,
    pipesCount: 0,
    flowing: null,
    ended: false,
    endEmitted: false,
    reading: false,
    sync: false,
    needReadable: false,
    emittedReadable: false,
    readableListening: false,
    resumeScheduled: false,
    emitClose: true,
    autoDestroy: false,
    destroyed: false,
    defaultEncoding: 'utf8',
    awaitDrain: 0,
    readingMore: false,
    decoder: null,
    encoding: null,
    [Symbol(kPaused)]: null
  },
  readable: true,
  _events: [Object: null prototype] {
    end: [Function: bound onceWrapper] { listener: [Function: onend] },
    prefinish: [Function: prefinish],
    finish: [ [Function], [Function] ],
    unpipe: [Function: onunpipe],
    error: [Function: onerror],
    close: [Function: bound onceWrapper] { listener: [Function: onclose] }
  },
  _eventsCount: 6,
  _maxListeners: undefined,
  _writableState: WritableState {
    objectMode: false,
    highWaterMark: 16384,
    finalCalled: false,
    needDrain: false,
    ending: false,
    ended: false,
    finished: false,
    destroyed: false,
    decodeStrings: true,
    defaultEncoding: 'utf8',
    length: 0,
    writing: false,
    corked: 0,
    sync: true,
    bufferProcessing: false,
    onwrite: [Function: bound onwrite],
    writecb: null,
    writelen: 0,
    afterWriteTickInfo: null,
    bufferedRequest: null,
    lastBufferedRequest: null,
    pendingcb: 0,
    prefinished: false,
    errorEmitted: false,
    emitClose: true,
    autoDestroy: false,
    bufferedRequestCount: 0,
    corkedRequestsFree: {
      next: null,
      entry: null,
      finish: [Function: bound onCorkedFinish]
    }
  },
  writable: true,
  allowHalfOpen: false,
  _transformState: {
    afterTransform: [Function: bound afterTransform],
    needTransform: false,
    transforming: false,
    writecb: null,
    writechunk: null,
    writeencoding: null
  },
  contentType: 'image/jpeg',
  filename: '14a8d90d-6149-408a-b6d3-1c32dec81d82.jpg',
  meta: {
    contentType: 'image/jpeg',
    filename: '14a8d90d-6149-408a-b6d3-1c32dec81d82.jpg'
  },
  [Symbol(kCapture)]: false
}

  */
